git/git_filter_protocol.go

```diff
@@ -0,0 +1,248 @@
+	}
+	_, err := o.w.WriteString(fmt.Sprintf("%04x", len(data)+4))
+	if err != nil {
+		return err
+	}
+	_, err = o.w.Write(data)
+	if err != nil {
+		return err
+	}
+	err = o.w.Flush()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *ObjectScanner) writeFlush() error {
+	_, err := o.w.WriteString(fmt.Sprintf("%04x", 0))
+	if err != nil {
+		return err
+	}
+	err = o.w.Flush()
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (o *ObjectScanner) writePacketText(data string) error {
+	//TODO: there is probably a more efficient way to do this. worth it?
+	return o.writePacket([]byte(data + "\n"))
+}
+
+func (o *ObjectScanner) writePacketList(list []string) error {
+	for _, i := range list {
+		err := o.writePacketText(i)
+		if err != nil {
+			return err
+		}
+	}
+	return o.writeFlush()
+}
+
+func (o *ObjectScanner) writeStatus(status string) error {
+	return o.writePacketList([]string{"status=" + status})
+}
+
+func (o *ObjectScanner) Init() bool {
+	tracerx.Printf("Initialize filter")
+	reqVer := "version=2"
+
+	initMsg, err := o.readPacketText()
+	if err != nil {
+		fmt.Fprintf(os.Stderr,
+			"Error: reading filter initialization failed with %s\n", err)
+		return false
+	}
+	if initMsg != "git-filter-client" {
+		fmt.Fprintf(os.Stderr,
+			"Error: invalid filter protocol welcome message: %s\n", initMsg)
+		return false
+	}
+
+	supVers, err := o.readPacketList()
+	if err != nil {
+		fmt.Fprintf(os.Stderr,
+			"Error: reading filter versions failed with %s\n", err)
+		return false
+	}
+	if !isStringInSlice(supVers, reqVer) {
+		fmt.Fprintf(os.Stderr,
+			"Error: filter '%s' not supported (your Git supports: %s)\n",
+			reqVer, supVers)
+		return false
+	}
+
+	err = o.writePacketList([]string{"git-filter-server", reqVer})
+	if err != nil {
+		fmt.Fprintf(os.Stderr,
+			"Error: writing filter initialization failed with %s\n", err)
+		return false
+	}
+	return true
+}
+
+func (o *ObjectScanner) NegotiateCapabilities() bool {
+	reqCaps := []string{"capability=clean", "capability=smudge"}
+
+	supCaps, err := o.readPacketList()
+	if err != nil {
+		fmt.Fprintf(os.Stderr,
+			"Error: reading filter capabilities failed with %s\n", err)
+		return false
+	}
+	for _, reqCap := range reqCaps {
+		if !isStringInSlice(supCaps, reqCap) {
+			fmt.Fprintf(os.Stderr,
+				"Error: filter '%s' not supported (your Git supports: %s)\n",
+				reqCap, supCaps)
+			return false
+		}
+	}
+
+	err = o.writePacketList(reqCaps)
+	if err != nil {
+		fmt.Fprintf(os.Stderr,
+			"Error: writing filter capabilities failed with %s\n", err)
+		return false
+	}
+
+	return true
+}
+
+func (o *ObjectScanner) ReadRequest() (map[string]string, []byte, error) {
+	tracerx.Printf("Process filter command.")
+
+	requestList, err := o.readPacketList()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	requestMap := make(map[string]string)
+	for _, pair := range requestList {
+		v := strings.Split(pair, "=")
+		requestMap[v[0]] = v[1]
+	}
```

Comment:
@peff I finally grokked your "parse as dictionary" comment completely. As you have already recognized it parses this nicely:
```
packet:          git> command=smudge
packet:          git> pathname=path/testfile.dat
```
But not this:
```
packet:          git> capability=clean
packet:          git> capability=smudge
```
... I wonder what you think about this?
```
packet:          git> clean=capability
packet:          git> smudge=capability
```
However, it's probably too late to change it. I guess in the end it is not that important and I don't want to annoy Junio with this kind of late change now that the code is in `next`.